File: java/src/P025/ReverseNodesinkGroup/ReverseNodesinkGroup.java
package P025.ReverseNodesinkGroup;

public class ReverseNodesinkGroup {
    public ListNode reverseKGroup(ListNode head, int k) {
        int count = 0;
        ListNode curr = head;
        // 获得k+1个节点
        while (curr != null && count != k) {
            curr = curr.next;
            count++;
        }
        if (count == k) {
            curr = this.reverseKGroup(curr, k);
            while (count-- > 0) {
                ListNode tmp = head.next;
                head.next = curr;
                curr = head;
                head = tmp;
            }
            head = curr;
        }

        return head;
    }
}

File: python/P025_Reverse_Nodes_in_k_Group/ReverseNodesinkGroup.py
# -*- coding: utf-8 -*-

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseKGroup(self, head, k):
        def reverse(first, last):
            prev = last
            while first != last:
                tmp = first.next
                first.next = prev
                prev, first = first, tmp
            return prev

        node = head
        for _ in range(k):
            if not node:
                # less than k, just leave the list as it is
                return head
            node = node.next
        new_head = reverse(head, node)
        head.next = self.reverseKGroup(node, k)
        return new_head